File: app/src/main/java/com/bakkenbaeck/sol/util/ThreeDayPhases.java
package com.bakkenbaeck.sol.util;


import android.location.Location;

import com.florianmski.suncalc.SunCalc;
import com.florianmski.suncalc.models.SunPhase;

import java.util.Calendar;
import java.util.Date;
import java.util.List;

public class ThreeDayPhases {

    public final static int NIGHT = 12;
    private final static int SUNRISE = 4;
    private final static int SUNSET = 8;

    private final Calendar today;
    private final Calendar tomorrow;
    private final Calendar yesterday;

    private List<SunPhase> yesterdaysSunPhases;
    private List<SunPhase> todaysSunPhases;
    private List<SunPhase> tomorrowsSunPhases;
    private long lastRefreshTime = 0;

    public ThreeDayPhases() {
        this.today = Calendar.getInstance();
        this.yesterday = Calendar.getInstance();
        this.tomorrow = Calendar.getInstance();

        this.yesterday.add(Calendar.DATE, -1);
        this.tomorrow.add(Calendar.DATE, 1);
    }

    public ThreeDayPhases init(final Location location) {
        if (!shouldRefresh()) {
            return this;
        }

        final double lat = location.getLatitude();
        final double lon = location.getLongitude();

        this.yesterdaysSunPhases = SunCalc.getPhases(this.yesterday, lat, lon);
        this.todaysSunPhases = SunCalc.getPhases(this.today, lat, lon);
        this.tomorrowsSunPhases = SunCalc.getPhases(this.tomorrow, lat, lon);

        return this;
    }

    public Calendar getDayLengthChangeBetweenTodayAndYesterday() {
        final Calendar todayLength = getDayLengthForPhases(this.todaysSunPhases);
        final Calendar yesterdayLength = getDayLengthForPhases(this.yesterdaysSunPhases);

        Calendar c = Calendar.getInstance();
        c.setTimeInMillis(todayLength.getTimeInMillis() - yesterdayLength.getTimeInMillis());

        return c;
    }

    public Calendar getDayLengthChangeBetweenTodayAndTomorrow() {
        final Calendar todayLength = getDayLengthForPhases(this.todaysSunPhases);
        final Calendar tomorrowLength = getDayLengthForPhases(this.tomorrowsSunPhases);

        Calendar c = Calendar.getInstance();
        c.setTimeInMillis(tomorrowLength.getTimeInMillis() - todayLength.getTimeInMillis());

        return c;
    }

    private Calendar getDayLengthForPhases(final List<SunPhase> sunPhases) {
        final SunPhase sunrise = sunPhases.get(SUNRISE);
        final SunPhase sunset = sunPhases.get(SUNSET);

        final Date sunriseTime = sunrise.getStartDate().getTime();
        final Date sunsetTime = sunset.getEndDate().getTime();
        final long dayLengthInMillis = sunsetTime.getTime() - sunriseTime.getTime();

        Calendar c = Calendar.getInstance();
        c.setTimeInMillis(dayLengthInMillis);

        return c;
    }

    public long getTomorrowsSunrise() {
        final SunPhase sunrise = this.tomorrowsSunPhases.get(SUNRISE);
        return sunrise.getStartDate().getTime().getTime();
    }

    public String getTodaysSunriseAsString() {
        final SunPhase sunrise = this.todaysSunPhases.get(SUNRISE);
        return DateUtil.dateFormat("HH:mm", sunrise.getStartDate().getTime());
    }

    public String getTodaysSunsetAsString() {
        final SunPhase sunset = this.todaysSunPhases.get(SUNSET);
        return DateUtil.dateFormat("HH:mm", sunset.getEndDate().getTime());
    }

    private boolean shouldRefresh() {
        final long refreshRate = 5000;
        final long nextRefreshThreshold = lastRefreshTime + refreshRate;
        final boolean shouldRefresh = System.currentTimeMillis() > nextRefreshThreshold;
        if (shouldRefresh) {
            lastRefreshTime = System.currentTimeMillis();
        }
        return shouldRefresh;
    }

    public CurrentPhase getCurrentPhase() {
        final Calendar now = Calendar.getInstance();
        for (final SunPhase phase : this.todaysSunPhases) {
            if (phase.getEndDate().after(now)) {
                return new CurrentPhase(phase);
            }
        }
        return new CurrentPhase(this.todaysSunPhases.get(0));
    }
}

File: app/src/main/java/com/bakkenbaeck/sol/util/ThreeDayPhases.kt
package com.bakkenbaeck.sol.util


import android.location.Location
import com.bakkenbaeck.sol.model.local.Phase

import com.florianmski.suncalc.SunCalc
import com.florianmski.suncalc.models.SunPhase

import java.util.Calendar
import java.util.Date

class ThreeDayPhases(location: Location) {

    companion object {
        private const val SUNRISE = 4
        private const val SUNSET = 8
        const val NIGHT = 12
    }

    private val today: Calendar by lazy { Calendar.getInstance() }
    private val tomorrow: Calendar by lazy { createCalendarWithDayOffset(1) }
    private val yesterday: Calendar by lazy { createCalendarWithDayOffset(-1) }

    private val yesterdaysSunPhases: List<SunPhase>
    private val todaysSunPhases: List<SunPhase>
    private val tomorrowsSunPhases: List<SunPhase>

    init {
        val lat = location.latitude
        val lon = location.longitude

        yesterdaysSunPhases = SunCalc.getPhases(yesterday, lat, lon)
        todaysSunPhases = SunCalc.getPhases(today, lat, lon)
        tomorrowsSunPhases = SunCalc.getPhases(tomorrow, lat, lon)
    }

    val dayLengthChangeBetweenTodayAndYesterday: Calendar
        get() {
            val todayLength = getDayLengthForPhases(todaysSunPhases)
            val yesterdayLength = getDayLengthForPhases(yesterdaysSunPhases)

            val c = Calendar.getInstance()
            c.timeInMillis = todayLength.timeInMillis - yesterdayLength.timeInMillis

            return c
        }

    val dayLengthChangeBetweenTodayAndTomorrow: Calendar
        get() {
            val todayLength = getDayLengthForPhases(todaysSunPhases)
            val tomorrowLength = getDayLengthForPhases(tomorrowsSunPhases)

            val c = Calendar.getInstance()
            c.timeInMillis = tomorrowLength.timeInMillis - todayLength.timeInMillis

            return c
        }

    val tomorrowsSunrise: Long
        get() {
            val sunrise = tomorrowsSunPhases[SUNRISE]
            return sunrise.startDate.time.time
        }

    val todaysSunriseAsLong: Long
        get() {
            val sunrise = todaysSunPhases[SUNRISE]
            return sunrise.startDate.timeInMillis
        }

    val todaysSunsetAsLong: Long
        get() {
            val sunset = todaysSunPhases[SUNSET]
            return sunset.endDate.timeInMillis
        }

    val currentPhase: Phase
        get() {
            val now = Calendar.getInstance()
            for (phase in todaysSunPhases) {
                if (phase.endDate.after(now)) {
                    return Phase(phase)
                }
            }
            return Phase(todaysSunPhases[0])
        }

    private fun createCalendarWithDayOffset(dayOffset: Int): Calendar {
        return Calendar.getInstance().apply { add(Calendar.DATE, dayOffset) }
    }

    private fun getDayLengthForPhases(sunPhases: List<SunPhase>): Calendar {
        val sunrise = sunPhases[SUNRISE]
        val sunset = sunPhases[SUNSET]

        val sunriseTime = sunrise.startDate.time
        val sunsetTime = sunset.endDate.time
        val dayLengthInMillis = sunsetTime.time - sunriseTime.time

        val c = Calendar.getInstance()
        c.timeInMillis = dayLengthInMillis

        return c
    }
}
